### DISCUSSION TITLE
Library linking problem on Linux

### DISCUSSION THREAD

**Discussion by SP-AWN** [IfcOpenShell/IfcOpenShell#5821] [Q&A] [answered]
Hi,
I'm trying to rewrite my old conversion tool which was using IfcOpenShell v0.6.0. I originally started with v0.8.0, but I ran into a bunch of problems, so I decided to switch to v0.7.0. It seems to work fine on Windows, but when I try to build my tool on Linux, I get undefined reference errors:
`/usr/bin/ld: /usr/local/lib/libIfcGeom.a(IteratorImplementation.cpp.o): in function `IteratorFactoryImplementation::IteratorFactoryImplementation()':
IteratorImplementation.cpp:(.text+0x2b1): undefined reference to `init_IteratorImplementation_Ifc2x3(IteratorFactoryImplementation*)'
/usr/bin/ld: IteratorImplementation.cpp:(.text+0x2b9): undefined reference to `init_IteratorImplementation_Ifc4(IteratorFactoryImplementation*)'
/usr/bin/ld: IteratorImplementation.cpp:(.text+0x2c1): undefined reference to `init_IteratorImplementation_Ifc4x1(IteratorFactoryImplementation*)'
/usr/bin/ld: IteratorImplementation.cpp:(.text+0x2c9): undefined reference to `init_IteratorImplementation_Ifc4x2(IteratorFactoryImplementation*)'
/usr/bin/ld: IteratorImplementation.cpp:(.text+0x2d1): undefined reference to `init_IteratorImplementation_Ifc4x3_rc1(IteratorFactoryImplementation*)'
/usr/bin/ld: IteratorImplementation.cpp:(.text+0x2d9): undefined reference to `init_IteratorImplementation_Ifc4x3_rc2(IteratorFactoryImplementation*)'
/usr/bin/ld: IteratorImplementation.cpp:(.text+0x2e1): undefined reference to `init_IteratorImplementation_Ifc4x3_rc3(IteratorFactoryImplementation*)'
/usr/bin/ld: IteratorImplementation.cpp:(.text+0x2e9): undefined reference to `init_IteratorImplementation_Ifc4x3_rc4(IteratorFactoryImplementation*)'
/usr/bin/ld: IteratorImplementation.cpp:(.text+0x2f1): undefined reference to `init_IteratorImplementation_Ifc4x3(IteratorFactoryImplementation*)'
/usr/bin/ld: IteratorImplementation.cpp:(.text+0x2f9): undefined reference to `init_IteratorImplementation_Ifc4x3_tc1(IteratorFactoryImplementation*)'
/usr/bin/ld: IteratorImplementation.cpp:(.text+0x301): undefined reference to `init_IteratorImplementation_Ifc4x3_add1(IteratorFactoryImplementation*)'
/usr/bin/ld: IteratorImplementation.cpp:(.text+0x309): undefined reference to `init_IteratorImplementation_Ifc4x3_add2(IteratorFactoryImplementation*)'
collect2: error: ld returned 1 exit status
`

My linking order:
target_link_libraries(fileconverter 
                        /usr/local/lib/libIfcParse.a
                        /usr/local/lib/libIfcGeom_ifc2x3.a
                        /usr/local/lib/libIfcGeom_ifc4.a
                        /usr/local/lib/libIfcGeom_ifc4x1.a
                        /usr/local/lib/libIfcGeom_ifc4x2.a
                        /usr/local/lib/libIfcGeom_ifc4x3_rc1.a
                        /usr/local/lib/libIfcGeom_ifc4x3_rc2.a
                        /usr/local/lib/libIfcGeom_ifc4x3_rc3.a
                        /usr/local/lib/libIfcGeom_ifc4x3_rc4.a
                        /usr/local/lib/libIfcGeom_ifc4x3.a
                        /usr/local/lib/libIfcGeom_ifc4x3_tc1.a
                        /usr/local/lib/libIfcGeom_ifc4x3_add1.a
                        /usr/local/lib/libIfcGeom_ifc4x3_add2.a
                        /usr/local/lib/libIfcGeom.a
                        /usr/local/lib/libSerializers_ifc2x3.a
                        /usr/local/lib/libSerializers_ifc4.a
                        /usr/local/lib/libSerializers_ifc4x1.a
                        /usr/local/lib/libSerializers_ifc4x2.a
                        /usr/local/lib/libSerializers_ifc4x3_rc1.a
                        /usr/local/lib/libSerializers_ifc4x3_rc2.a
                        /usr/local/lib/libSerializers_ifc4x3_rc3.a
                        /usr/local/lib/libSerializers_ifc4x3_rc4.a
                        /usr/local/lib/libSerializers_ifc4x3.a
                        /usr/local/lib/libSerializers_ifc4x3_tc1.a
                        /usr/local/lib/libSerializers_ifc4x3_add1.a
                        /usr/local/lib/libSerializers_ifc4x3_add2.a
                        /usr/local/lib/libSerializers.a
)
What I'm doing wrong?

**Comment by aothms** [comment, marked as answer]
Keep in mind that the linux linker is not recursive (unlike the one on window as you observed). 

So the order is significant.

`init_IteratorImplementation_Ifc2x3` is defined in `libIfcGeom_ifc2x3` and referenced in `libIfcGeom`, but in your linker order `libIfcGeom_ifc2x3` is before `libIfcGeom`, so when it encounters the imported symbol, it has already forgotten it.

Honestly, the easiest solution is to just take that list and duplicate it a couple of time because there could be cyclic dependencies among them.

You could also use --start-group --end-group to make `ld` cycle through the list.

Coincidentally I just implemented ifopsh08 compatibility in the voxelization_toolkit, might be of some help. https://github.com/IfcOpenShell/voxelization_toolkit/commit/c563c0df0dadbd8d8d28e80cb41f3a3556d6f9f7

But yes, I fully acknowledge that interfacing with IfcOpenShell in C++ is much harder then I would want it to be. There should really just be a FindIfcOpenShell cmake module that people can use, but our plates are always too full.

All help is welcome :)

**Reply by SP-AWN** [reply]
Thanks for help
this is final solution thats seems to be working:
target_link_libraries(fileconverter 
                        -Wl,--start-group
                        /usr/local/lib/libIfcParse.a
                        /usr/local/lib/libIfcGeom_ifc2x3.a
                        /usr/local/lib/libIfcGeom_ifc4.a
                        /usr/local/lib/libIfcGeom_ifc4x1.a
                        /usr/local/lib/libIfcGeom_ifc4x2.a
                        /usr/local/lib/libIfcGeom_ifc4x3_rc1.a
                        /usr/local/lib/libIfcGeom_ifc4x3_rc2.a
                        /usr/local/lib/libIfcGeom_ifc4x3_rc3.a
                        /usr/local/lib/libIfcGeom_ifc4x3_rc4.a
                        /usr/local/lib/libIfcGeom_ifc4x3.a
                        /usr/local/lib/libIfcGeom_ifc4x3_tc1.a
                        /usr/local/lib/libIfcGeom_ifc4x3_add1.a
                        /usr/local/lib/libIfcGeom_ifc4x3_add2.a
                        /usr/local/lib/libIfcGeom.a
                        /usr/local/lib/libIfcParse.a
                        /usr/local/lib/libSerializers_ifc2x3.a
                        /usr/local/lib/libSerializers_ifc4.a
                        /usr/local/lib/libSerializers_ifc4x1.a
                        /usr/local/lib/libSerializers_ifc4x2.a
                        /usr/local/lib/libSerializers_ifc4x3_rc1.a
                        /usr/local/lib/libSerializers_ifc4x3_rc2.a
                        /usr/local/lib/libSerializers_ifc4x3_rc3.a
                        /usr/local/lib/libSerializers_ifc4x3_rc4.a
                        /usr/local/lib/libSerializers_ifc4x3.a
                        /usr/local/lib/libSerializers_ifc4x3_tc1.a
                        /usr/local/lib/libSerializers_ifc4x3_add1.a
                        /usr/local/lib/libSerializers_ifc4x3_add2.a
                        /usr/local/lib/libSerializers.a
                        -Wl,--end-group
)
